### PR TITLE
Decimal.GetHashCode tests

### PR DESCRIPTION
**Pull request by pentp**
This adds tests for Decimal.GetHashCode to detect situations where the same number gets different hashcodes if the scaling is different (https://github.com/dotnet/coreclr/issues/2944).

These tests will fail until https://github.com/dotnet/corert/pull/5886 is merged into CoreCLR and CoreFX.

### CODE REVIEW

File: src/System.Runtime/tests/System/DecimalTests.cs [hunk 1]
```diff
@@ -1602,7 +1602,23 @@ public static void BigInteger_RoundAwayFromZero()
             }
         }
 
-        static decimal[] GetRandomData(out BigDecimal[] bigDecimals)
+        [Fact]
+        public static new void GetHashCode()
+        {
+            var dict = new Dictionary<string, (int hash, string value)>();
+            foreach (var d in GetRandomData(out _, hash: true))
```
Comment (stephentoub) [inline src/System.Runtime/tests/System/DecimalTests.cs]: Nit: var => decimal

File: src/System.Runtime/tests/System/DecimalTests.cs [hunk 2]
```diff
@@ -1602,7 +1602,23 @@ public static void BigInteger_RoundAwayFromZero()
...
+            {
+                var value = d.ToString(CultureInfo.InvariantCulture);
```
Comment (stephentoub) [inline src/System.Runtime/tests/System/DecimalTests.cs]: nit: var => string

File: src/System.Runtime/tests/System/DecimalTests.cs [hunk 3]
```diff
@@ -1602,7 +1602,23 @@ public static void BigInteger_RoundAwayFromZero()
...
+                var key = value[value.Length - 1] == '0' && value.Contains('.') ? value.AsSpan().TrimEnd('0').TrimEnd('.').ToString() : value;
```
Comment (stephentoub) [inline src/System.Runtime/tests/System/DecimalTests.cs]: Nit: var => string

File: src/System.Runtime/tests/System/DecimalTests.cs [hunk 4]
```diff
@@ -1602,7 +1602,23 @@ public static void BigInteger_RoundAwayFromZero()
...
+                var hash = d.GetHashCode();
```
Comment (stephentoub) [inline src/System.Runtime/tests/System/DecimalTests.cs]: Nit: var => int

File: src/System.Runtime/tests/System/DecimalTests.cs [hunk 5]
```diff
@@ -1621,9 +1637,26 @@ static decimal[] GetRandomData(out BigDecimal[] bigDecimals)
...
+                                    d *= 1.0m;
+                                    unsafe
+                                    {
+                                        var curScale = (byte)(*(uint*)&d >> BigDecimal.ScaleShift);
+                                        if (curScale <= lastScale)
+                                            break;
+                                        lastScale = curScale;
+                                    }
+                                    list.Add(d);
+                                }
```
Comment (stephentoub) [inline src/System.Runtime/tests/System/DecimalTests.cs]: Does this significantly impact the running time of the tests?

File: src/System.Runtime/tests/System/DecimalTests.cs [hunk 6]
```diff
@@ -1602,7 +1602,23 @@ public static void BigInteger_RoundAwayFromZero()
...
+            var dict = new Dictionary<string, (int hash, string value)>();
+            foreach (var d in GetRandomData(out _, hash: true))
+            {
+                var value = d.ToString(CultureInfo.InvariantCulture);
+                var key = value[value.Length - 1] == '0' && value.Contains('.') ? value.AsSpan().TrimEnd('0').TrimEnd('.').ToString() : value;
+                var hash = d.GetHashCode();
+                if (!dict.TryGetValue(key, out var ex))
+                    dict.Add(key, (hash, value));
+                else if (ex.hash != hash)
+                    throw new Xunit.Sdk.XunitException($"Decimal {key} has multiple hash codes: {ex.hash} ({ex.value}) and {hash} ({value})");
```
Comment (stephentoub) [inline src/System.Runtime/tests/System/DecimalTests.cs]:
```C#
else
{
    Assert.True(ex.hash == hash, $"Decimal {key} ... ");
}
```
? Please also use braces for both the if and the else branches.

File: src/System.Runtime/tests/System/DecimalTests.cs (hunk 5 above)
Comment (pentp) [reply src/System.Runtime/tests/System/DecimalTests.cs]: No, it takes 6ms to run the GetHashCode tests and other tests aren't affected.

Comment (stephentoub) [reply src/System.Runtime/tests/System/DecimalTests.cs]: Cool, thanks.

**Comment by jkotas** [comment]
Waiting for CoreCLR update

**Comment by jkotas** [comment]
@dotnet-bot test this please

**Comment by jkotas** [comment]
@pentp Could you please add `[SkipOnTargetFramework(TargetFrameworkMonikers.NetFramework, "Full framework does not have the fix for this bug")]` attribute on the new test to make the CI green?

**Pull request merged**